### PR TITLE
Blob txs reorgs

### PR DESCRIPTION
**Pull request by marcindsobczak**
Opened as continuation of https://github.com/NethermindEth/nethermind/pull/5954

## Changes

- add to `TxPoolConfig` option of enabling support for blob txs reorgs `BlobReorgsSupportEnabled`, `true` by default
- add new column `ProcessedTxs` to `BlobTransactions`  db dedicated for processed blob transactions
- after inclusion of blob tx, it is removed from standard mempool/db. All txs from processed block are added to processed txs db where `blockNumber` is a key and value is array of txs
- in case of reorg, if we have in db txs from reorged block, they are readded to standard mempool/db and removed from processed db
- add `ProcessedTransactionsDbCleaner` which is triggered by `ForkChoiceUpdated` and is removing all KeyValuePairs with blockNumber not higher than finalized block, as reorg is not possible after finalization

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.TxPool/BlobTxStorage.cs [hunk 1]
```diff
@@ -81,6 +85,50 @@ public void Delete(in ValueHash256 hash, in UInt256 timestamp)
...
+
+        int contentLength = 0;
+        foreach (Transaction transaction in blockBlobTransactions)
+        {
+            contentLength += _txDecoder.GetLength(transaction, RlpBehaviors.InMempoolForm);
+        }
+
+        IByteBuffer byteBuffer = PooledByteBufferAllocator.Default.Buffer(Rlp.LengthOfSequence(contentLength));
+        using NettyRlpStream rlpStream = new(byteBuffer);
+        rlpStream.StartSequence(contentLength);
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.TxPool/BlobTxStorage.cs:103]: Cosmetic proposition: extract RLP encoding/decoding to separate methods

File: src/Nethermind/Nethermind.TxPool/TxPoolConfig.cs [hunk 2]
```diff
@@ -14,6 +14,7 @@ public class TxPoolConfig : ITxPoolConfig
                                                                         // every day about 21600 blobs will be included (7200 blocks per day * 3 blob target)
         public int BlobCacheSize { get; set; } = 256;
         public int InMemoryBlobPoolSize { get; set; } = 512; // it is used when persistent pool is disabled
+        public bool BlobReorgsSupportEnabled { get; set; } = true;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.TxPool/TxPoolConfig.cs:17]: What if we merge it even now, but with this flag false for now?

File: src/Nethermind/Nethermind.TxPool/TxPool.cs [hunk 3]
```diff
@@ -248,31 +272,47 @@ private void RemoveProcessedTransactions(Transaction[] blockTransactions)
...
+                    blobTxs++;
+                    blobs += blockTx.BlobVersionedHashes?.Length ?? 0;
+
+                    if (_blobReorgsSupportEnabled)
+                    {
+                        if (_blobTransactions.TryGetValue(blockTx.Hash, out Transaction? fullBlobTx))
+                        {
+                            if (_logger.IsTrace) _logger.Trace($"Saved processed blob tx {blockTx.Hash} from block {block.Number} to ProcessedTxs db");
+                            blobTxsToSave ??= new List<Transaction>(Eip4844Constants.GetMaxBlobsPerBlock());
+                            blobTxsToSave.Add(fullBlobTx);
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs:294]: So I think you could remove the need of "blobTxsToSave" list with saving to db and db batches

File: src/Nethermind/Nethermind.Merge.Plugin.Test/ProcessedTransactionsDbCleanerTests.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        returnedTxs!.Length.Should().Be(2);
+
+        IBlockFinalizationManager finalizationManager = Substitute.For<IBlockFinalizationManager>();
+        ProcessedTransactionsDbCleaner dbCleaner = new(finalizationManager, columnsDb.GetColumnDb(BlobTxsColumns.ProcessedTxs), _logManager);
+
+        finalizationManager.BlocksFinalized += Raise.EventWith(
+            new FinalizeEventArgs(Build.A.BlockHeader.TestObject,
+                Build.A.BlockHeader.WithNumber(finalizedBlock).TestObject));
+
+        Thread.Sleep(100);
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin.Test/ProcessedTransactionsDbCleanerTests.cs:57]: Why Thread.Sleep is needed here? Let's not use Thread.Sleep in tests because it is recipe for flaky tests

File: src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            Task.Run(() => CleanProcessedTransactionsDb(e.FinalizedBlocks[0].Number));
+        }
+    }
+
+    private void CleanProcessedTransactionsDb(long newlyFinalizedBlockNumber)
+    {
+        try
+        {
+            List<long> blocksInDb = new();
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs:43]: You can remove this list quite easily, do it before @LukaszRozmej review :D

File: src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        _finalizationManager = finalizationManager ?? throw new ArgumentNullException(nameof(finalizationManager));
+        _processedTxsDb = processedTxsDb ?? throw new ArgumentNullException(nameof(processedTxsDb));
+        _logger = logManager.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+
+        _finalizationManager.BlocksFinalized += OnBlocksFinalized;
+    }
+
+    private void OnBlocksFinalized(object? sender, FinalizeEventArgs e)
+    {
+        if (e.FinalizedBlocks.Count > 0 && e.FinalizedBlocks[0].Number > _lastFinalizedBlock)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs:33]: In scenario of the merge plugin this e.FinalizedBlocks won't be null, but something to consider if null check would be good here

File: src/Nethermind/Nethermind.TxPool.Test/TxPoolTests.Blobs.cs [hunk 7]
```diff
@@ -490,5 +491,61 @@ public void should_calculate_size_of_blob_tx_correctly(int numberOfBlobs, int ex
                 .TestObject;
             blobTx.GetLength().Should().Be(expectedLength);
         }
+
+        [Test]
+        public async Task should_add_processed_txs_to_db()
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.TxPool.Test/TxPoolTests.Blobs.cs:496]: no test scenarios for block reorgs? I think if reorgs are super hard to test in devnets it would be nice to add unit tests for a few reogrs scenarios

File: src/Nethermind/Nethermind.TxPool/ITxStorage.cs [hunk 8]
```diff
@@ -15,4 +15,7 @@ public interface ITxStorage
     IEnumerable<LightTransaction> GetAll();
     void Add(Transaction transaction);
     void Delete(in ValueHash256 hash, in UInt256 timestamp);
+    bool TryGetBlobTransactionsFromBlock(long blockNumber, out Transaction[]? blockBlobTransactions);
+    void AddBlobTransactionsFromBlock(long blockNumber, IList<Transaction> blockBlobTransactions);
+    void DeleteBlobTransactionsFromBlock(long blockNumber);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/ITxStorage.cs:20]: Maybe we should extract it to `IBlobTxStorage` interface? Even if same class would implement both (maybe it should be separate class?)

File: src/Nethermind/Nethermind.TxPool/TxPool.cs (hunk 3 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:294]: If not you could easily change it to lazy evaluated `IEnumerable` or at least `ArrayPoolList`, but batch could be even simpler.

File: src/Nethermind/Nethermind.TxPool/TxPoolConfig.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool/TxPoolConfig.cs:17]: Should we make the actual logic based on this one && if chain supports blobs (or will support in the future)

File: src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs (hunk 6 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs:33]: ok, resolved this

File: src/Nethermind/Nethermind.TxPool/TxPoolConfig.cs (hunk 2 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.TxPool/TxPoolConfig.cs:17]: It is already used only when all `BlobReorgsSupportEnabled`, `BlobSupportEnabled` and `PersistentBlobStorageEnabled` are true, so as blob support and persistent storage are false for default now, we can merge with reorgs = true, it will be disabled anyway
https://github.com/NethermindEth/nethermind/blob/dcbb3d35d490dab70f972025dc57d051b0230271/src/Nethermind/Nethermind.TxPool/TxPool.cs#L97

File: src/Nethermind/Nethermind.TxPool/TxPool.cs (hunk 3 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:294]: It works this way - collecting txs, then creating one long RLP with all of them and then saving to db as one record, where block number is a key and all blob txs from this block (as RLP) are a value. So we need to collect all txs before encoding - otherwise we don't know message length which is a prerequisit for encoding. As we need to collect all txs at some point, lazy enumeration will just move this part to BlobTxStorage, but we will not avoid it. Doing batch write will not help here, as we are doing only one write, unless I am missing something. I just changed `blobTxsToSave` to be ArrayPoolList

File: src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+            if (_logger.IsTrace) _logger.Trace($"There are processed blob txs from {blocksInDb.Count} blocks in ProcessedTxs db");
+            foreach (long blockNumber in blocksInDb)
+            {
+                if (newlyFinalizedBlockNumber >= blockNumber)
+                {
+                    if (_logger.IsTrace) _logger.Trace($"Cleaning processed blob txs from block {blockNumber}");
+                    _processedTxsDb.Delete(blockNumber);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs]: Without list it will be problematic when cleaning first time (iterating from block 0 to 18.5M on mainnet) and after every restart. At any point we should have only up to ~100 blocks in db so it won't be expensive to once per epoch just read all keys, keep in memory list of ~100 block numbers and delate only actually existing data

I changed delating from db to be done in batch

File: src/Nethermind/Nethermind.TxPool/ITxStorage.cs (hunk 8 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.TxPool/ITxStorage.cs]: We are storing only blob txs for now, doesn't look like we will need to store other types in the future, but if we will - no problem with separating it. I created `IBlobTxStorage` extending `ITxStorage` and moved blob-specific methods

File: src/Nethermind/Nethermind.TxPool/TxPoolConfig.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool/TxPoolConfig.cs:18]: Two things:
1. Maybe better to collapse all those 3 config items into one enum like:

`BlobsSupport: { Disabled, InMemory, Storage, StorageWithReorgs } ` ?

2. Should we also query chainspec to see if 4844 is ever enabled on it and if not disable blobs?

File: src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs (hunk 5 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs:43]: Yeah, you probably don't need it, you can iterate and remove in the batch directly

File: src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs]: yes but you don't need the list, you can just delete things in the batch directly

File: src/Nethermind/Nethermind.TxPool/IBlobTxStorage.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Collections.Generic;
+using Nethermind.Core;
+
+namespace Nethermind.TxPool;
+
+public interface IBlobTxStorage : ITxStorage
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/IBlobTxStorage.cs:9]: do we need this inheritance, can't they be separated?

File: src/Nethermind/Nethermind.TxPool.Test/TxPoolTests.Blobs.cs [hunk 11]
```diff
@@ -494,14 +495,128 @@ public void should_calculate_size_of_blob_tx_correctly(int numberOfBlobs, int ex
...
+                .Excluding(t => t.GasBottleneck)    // GasBottleneck is not encoded/decoded...
+                .Excluding(t => t.PoolIndex)        // ...as well as PoolIndex
+                .Excluding(t => t.SenderAddress));  // sender is recovered later, it is not returned from db
+
+            blobTxStorage.DeleteBlobTransactionsFromBlock(blockNumber);
+            blobTxStorage.TryGetBlobTransactionsFromBlock(blockNumber, out returnedTxs).Should().BeFalse();
+        }
+
+        [Test]
+        public async Task should_bring_back_reorganized_blob_txs()
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool.Test/TxPoolTests.Blobs.cs:549]: hard to follow tests, can they be split to simpler steps?

**Review by MarekM25 — APPROVED** [review]

**Review by LukaszRozmej — APPROVED** [review]
